src/main.rs

```diff
@@ -172,57 +199,56 @@ struct PatchQuery {
     erase: Vec<String>,
 }
 
-async fn patch(uid: i64, q: PatchQuery, pool: DB) -> impl Reply {
-    // todo: sane error handling
-
+async fn patch_signals(pool: &DB, uid: i64, q: PatchQuery) -> eyre::Result<()> {
     for tag in q.add {
         println!("add {}", &tag);
-        sqlx::query(
-            "
-insert into signal (user_id, url, tag, signal)
-values ($1, $2, $3, $4)
-on conflict (user_id, url, tag) do update set signal = $4
-        ",
-        )
-        .bind(uid)
-        .bind(&q.url)
-        .bind(tag)
-        .bind(true)
-        .execute(&pool)
-        .await
-        .unwrap();
+        Signal::set(pool, uid, &q.url, &tag, true)
+            .await
+            .wrap_err("failed to add signal")?
     }
 
     for tag in q.rm {
         println!("rm {}", &tag);
-        sqlx::query(
-            "
-insert into signal (user_id, url, tag, signal)
-values ($1, $2, $3, $4)
-on conflict (user_id, url, tag) do update set signal = $4
-        ",
-        )
-        .bind(uid)
-        .bind(&q.url)
-        .bind(tag)
-        .bind(false)
-        .execute(&pool)
-        .await
-        .unwrap();
+        Signal::set(pool, uid, &q.url, &tag, false)
+            .await
+            .wrap_err("failed to rm signal")?
     }
 
     for tag in q.erase {
         println!("erase {}", &tag);
-        sqlx::query("delete from signal where user_id = $1 and url = $2 and tag = $3")
-            .bind(uid)
-            .bind(&q.url)
-            .bind(tag)
-            .execute(&pool)
+        Signal::erase(pool, uid, &q.url, &tag)
             .await
-            .unwrap();
+            .wrap_err("failed to erase signal")?
     }
 
     println!();
+    Ok(())
+}
 
-    warp::reply::reply()
+async fn patch(uid: i64, q: PatchQuery, pool: DB) -> http::Response<hyper::Body> {
```

Comment:
Two functions `patch` and `patch_signals` with reordered parameters will be *so* confusing to maintain...

Reply:
I think http verb named functions were doomed from the start :p I started making them more specific and moving them to different namespaces in some later commits, definitely need a proper pass at cleaning it up.

The ordering is my bad -- think I normally tend to write more general args first and go towards more specific which is the opposite of what the initial verb-functions had. I can loop around and clean up my ordering. What is your style here? How would you order client vs pool?

Reply:
I usually tend to put the more specific arguments first, because after I read (or write) the name of the function (especially an overloaded one), I usually care more about what distinguishes it from the others. But the specific order doesn't matter too much to me here, I just think that we should pick one and stick to it everywhere :)

Reply:
Fixed by moving `pool` to the end here.